### DISCUSSION TITLE
Providers which are internal to the component and are not exported

### DISCUSSION THREAD

**Discussion by dimsuz** [evant/kotlin-inject#260] [Ideas]
Hi! Thanks for the wonderful library!

I wonder how hard would it be to have some providers allowed to be "internal", for example

```kotlin
@Component
abstract class NetworkComponent {
  @Provides
  protected fun provideOkHttpClient(): OkHttpClient = { /* ... */ }
  
  abstract val repo: Repo
}

@Inject
class Repo(val httpClient: OkHttpClient)

// Not possible to do this, will complain about "protected" provides-function above
@Component
abstract class DomainComponent(@Component val parent: NetworkComponent)
```

I.e. in my setup (perhaps a multi-gradle-module one) I only want the component to "export" `Repo` and hide `OkHttpClient`, which I want to provide only "internally", because while building it I still might want to use DI, as it could be helpful if my dependency building is complex, but in the end I don't want anything apart the final object to "leak" and be visible/usable on the consuming side. Trying to inject `OkHttpClient` in any child component would be an error in this case.

This is what I very much miss in Dagger, perhaps `kotlin-inject` could have such a feature?

**Comment by evant** [comment]
You can do this today by hiding your component behind an interface.

```kotlin
interface NetworkComponent {
   val repo: Repo
}

@Component
abstract class NetworkComponentImpl : NetworkComponent {
  @Provides
  fun provideOkHttpClient(): OkHttpClient = { /* ... */ }
}

@Inject
class Repo(val httpClient: OkHttpClient)

@Component
abstract class DomainComponent(@Component val parent: NetworkComponent)
```

but yeah there could be room for improvement here. There tricky part is knowing what to generate so that the transitive dependencies can still be constructed in a downstream component, which could be in another module.

**Reply by dimsuz** [reply]
Wow, that's great, I will try this!

Yes, I also thought that this could be tricky to implement, library would have to make quite a few deductions, oh and across module boundary this could be even trickier, not sure what ksp would "see" in this case (I only tried to toy with it).
